server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -99,12 +100,13 @@ public MappedFieldType(String name, boolean isIndexed, boolean hasDocValues, Tex
      * Return a fielddata builder for this field
      *
      * @param fullyQualifiedIndexName the name of the index this field-data is build for
+     * @param context holds information that may be needed when loading fielddata
      *
      * @throws IllegalArgumentException if the fielddata is not supported on this type.
      * An IllegalArgumentException is needed in order to return an http error 400
      * when this error occurs in a request. see: {@link org.elasticsearch.ExceptionsHelper#status}
      */
-    public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+    public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, FieldDataContext context) {
```

Comment:
It could be a `Supplier<SearchLookup>` ? Why not pass the `SearchLookup` directly or `null` if not applicable (for index sorting for instance) ?

Reply:
I also wonder how we can protect against self-reference ? It  should  be forbidden  to call `getForField` on the same field name since this would create an infinite loop ? Maybe a wrapper around SearchLookup is enough but this would only detect it at runtime, curious to hear @nik9000 thoughts on this

Reply:
agreed it could just be a supplier. I tried the context idea because it gives room to add more in the future, and also because I was initially thinking of just an interface and make query shard context implement it but I changed approach on the way. I think supplier would be fine as well and we can always change it if we need something different. On the other hand, there was another recent request to pass shard info through, so maybe we should take that one into account? see #57231 .

One alternative we could consider, that is less invasive but also less visible, would be to not change the signature of fielddataBuilder and rather augment the fielddata that gets loaded only when needed.

Recursion is indeed something to think about. I am not sure if it can be checked at compile time and whether such check belongs in this PR, it could also be up to the field data impl once the script is known to check for self references.

Reply:
I'm not sure we know enough about how this'll be used to know how to deal with recursive loops. 

I do like the idea of a wrapper to detect the loops though. That was other parts of the code don't have to deal with the overhead. I'm fairly sure that we can't know up front what fields a script will try to access without some *interesting* changes either in the script infrastructure or the way users write scripts or both. But I don't think we know enough yet. I'd certainly be willing to play with it some in the next few days though.

I'm comfortable adding this without any explicit loop detection for now and figuring that out as we go.

Also, I would have suggested the `Supplier` route, but given #57231 maybe we should just add the `int shardId` to the interface? It won't be functional any more, but that isn't too bad.

Reply:
Maybe I am concerned because the pr targets master. I am totally fine leaving the detect recursive loop to later as long as this code lives in a branch in the meantime. @javanna why not targeting the scripted field branch ?

Reply:
I like passing the lookup here because *some* fields will be able to know up front which fields they need so it'd be nice to be able to snag them early. Or at least validate that they are snag-able.

Reply:
> why not targeting the scripted field branch

Ideally the branch has changes only related to the plugin that it adds. I thought that the need is clear and I am not so keen on having the branch grow big in terms of infra that is needed to wire dependencies. In short, we could have hacked this in the branch but I prefer to discuss and tackle this specific problem sooner rather than later. I also did not think that we would want to tackle recursion straight-away, but if we are seeing that we need to know more before we make this change, we can make it to the feature branch for now.

Reply:
> I like passing the lookup here because some fields will be able to know up front which fields they need so it'd be nice to be able to snag them early. Or at least validate that they are snag-able.

@nik9000 I don't follow, can you expand?

Reply:
> @nik9000 I don't follow, can you expand?

Sure! I think that script-based fields won't know up front how they are going to use the lookup, but that a, say, a runtime field that runs an ml model *will* know which fields it needs from the `SearchLookup`. It could call `getDoc().getForField(blah)` up front, even failing if the field that it needs doesn't exist.